### PR TITLE
[MoE][PyTorch] Add mask-based MoE permutation

### PR DESCRIPTION
**Pull request by hxbai**
# Description

Add mask-based token permutation and local chunk permutation fused kernels. These kernels are implemented with OpenAI Triton.

Related commit in Megatron-LM https://github.com/NVIDIA/Megatron-LM/commit/ac0474d26e8d8688b950704ea58d14858c9d047f

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Non-breaking API changes in `te.pytorch.permutation.moe_permute` and `te.pytorch.permutation.moe_unpermute`
- Add new APIs of `te.pytorch.permutation.moe_sort_chunks_by_indices`

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/permutation.py [hunk 1]
```diff
@@ -18,8 +19,8 @@
 ]
```
Comment (yaox12) [inline transformer_engine/pytorch/permutation.py:19]: Add `moe_chunk_permute` and `moe_chunk_unpermute` here.

File: transformer_engine/pytorch/permutation.py [hunk 2]
```diff
@@ -2,24 +2,26 @@
...
-"""Linear API"""
+"""MoE Permutaion API"""
 import warnings
 from typing import Tuple
 import torch
 
 import transformer_engine_torch as tex
-from .constants import TE_DType
-from .float8_tensor import Float8Tensor
+import transformer_engine.pytorch.triton.permutation as triton_permuataion
```
Comment (timmoon10) [inline transformer_engine/pytorch/permutation.py:11]: Nit:
```suggestion
import transformer_engine.pytorch.triton.permutation as triton_permutation
```

File: transformer_engine/pytorch/permutation.py [hunk 3]
```diff
@@ -225,36 +227,268 @@ def backward(
...
+        if not permuted_act_grad.numel():
+            return permuted_act_grad, None, None
+
+        act_grad = None
+        if ctx.needs_input_grad[0]:
+            (row_id_map,) = ctx.saved_tensors
+            if ctx.fp8:
+                assert isinstance(
+                    permuted_act_grad, Float8Tensor
+                ), "Grad of the output must be in Float8Tensor type for FP8 moe_permute."
```
Comment (timmoon10) [inline transformer_engine/pytorch/permutation.py:295]: Couldn't we decouple FP8 in the forward and backward?
```suggestion
            fp8 = isinstance(permuted_act_grad, Float8Tensor)
            if fp8:
```
If there are no obstacles, we could also do the same thing for `_moe_unpermute_mask_map` and `_moe_chunk_sort`.

File: tests/pytorch/test_permutation.py [hunk 4]
```diff
@@ -362,6 +455,414 @@ def backward_wrapper(
...
+        te_unpermute_fwd_input, row_id_map, te_probs, restore_shape, map_type="mask"
+    )
+    te_unpermute_output.backward(te_unpermute_bwd_input, retain_graph=True)
+
+    ###################################################################################################################################
+    #
+    # Results Check
+    #
+    ###################################################################################################################################
+    tols = dtype_tols(te_dtype)
```
Comment (timmoon10) [inline tests/pytorch/test_permutation.py:588]: Shouldn't we expect bit-wise exact results?
```suggestion
    tols = { "atol": 0, "rtol": 0 }
```

File: tests/pytorch/test_permutation.py [hunk 5]
```diff
@@ -362,6 +455,414 @@ def backward_wrapper(
...
+
+    te_output = te_sort_chunks_by_indices(te_fwd_input, split_sizes_cuda, sorted_idxs_cuda)
+    te_output.backward(te_bwd_input, retain_graph=True)
+
+    ###################################################################################################################################
+    #
+    # Results Check
+    #
+    ###################################################################################################################################
+    tols = dtype_tols(te_dtype)
```
Comment (timmoon10) [inline tests/pytorch/test_permutation.py:797]: We should expect bit-wise exact results.
```suggestion
    tols = { "atol": 0, "rtol": 0 }
```

File: transformer_engine/pytorch/triton/permutation.py [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+):
+    pid_m = tl.program_id(0)
+    pid_n = tl.program_id(1)
+    offset = pid_n * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    expert_token_mask = tl.load(
+        routing_map_ptr + pid_m * stride_routing_map_expert + offset * stride_routing_map_token,
+        mask=(offset < num_tokens),
+        other=0,
+    ).to(tl.int64)
+    expert_token_cumsum = tl.cumsum(expert_token_mask) * expert_token_mask
```
Comment (phu0ngng) [inline transformer_engine/pytorch/triton/permutation.py:38]: An interesting way to exclude the zero token_mask. Happy to learn!

File: transformer_engine/pytorch/triton/permutation.py [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+    pid_n = tl.program_id(1)
+    chunk_idx = pid_m * tl.cdiv(num_tokens, BLOCK_SIZE) + pid_n
+    offset = pid_n * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    chunk_cumsum = tl.load(
+        row_id_map_ptr + pid_m * num_tokens + offset, mask=(offset < num_tokens), other=0
+    )
+
+    workspace_off = tl.arange(0, WORKSPACE_LOAD_WIDTH)
+    chunk_sums = tl.load(workspace_ptr + workspace_off, mask=workspace_off < chunk_idx)
+    chunk_cumsum = tl.where(chunk_cumsum == 0, -1, chunk_cumsum + tl.sum(chunk_sums) - 1)
```
Comment (phu0ngng) [inline transformer_engine/pytorch/triton/permutation.py:67]: These three names `chuck_cumsum`, `chuck_sums`, and `chunk_cumsum` are quite confusing.
If I understand it correctly, I suggest to rename them to:
- `chuck_cumsum` -> `row_id_map_within_block`
- `chuck_sums` -> n_tokens_per_expert
- `chuck_cumsum` -> `row_id_map_within_grid`

In addition, I think we should move the `-1` to the `pass1` as it is the correction for the calculation of `expert_token_cumsum`, as:
```
expert_token_cumsum = (tl.cumsum(expert_token_mask) - 1) * expert_token_mask
```

File: transformer_engine/pytorch/permutation.py [hunk 8]
```diff
@@ -2,24 +2,26 @@
...
 __all__ = [
     "moe_permute",
     "moe_unpermute",
+    "moe_sort_chunks_by_indices",
 ]
 
 
-class _moe_permute(torch.autograd.Function):
-    """functional Permute"""
+class _moe_permute_indice_map(torch.autograd.Function):
```
[the remark this answers is not in the document]
Comment (hxbai) [reply transformer_engine/pytorch/permutation.py]: OK, modified.

File: transformer_engine/pytorch/permutation.py (hunk 2 above)
Comment (hxbai) [reply transformer_engine/pytorch/permutation.py]: Fixed.

File: transformer_engine/pytorch/permutation.py (hunk 3 above)
Comment (hxbai) [reply transformer_engine/pytorch/permutation.py]: Modified. Now for bwd, it would follow the dtype of the grad tensor.

File: tests/pytorch/test_permutation.py (hunk 5 above)
Comment (hxbai) [reply tests/pytorch/test_permutation.py]: Like the one above, I changed to bit-wise matching except for fp8.

File: transformer_engine/pytorch/triton/permutation.py (hunk 7 above)
Comment (hxbai) [reply transformer_engine/pytorch/triton/permutation.py]: Thanks. You are right. I modified these names (renamed `chunk_sums` to `n_tokens_per_block` rather than `n_tokens_per_expert`).

For the `-1`, if we move it to `pass1`, then we cannot easily distinguish the `row_id: 0` and the `mask: 0` and we need extra ways to handle whether it is masked out. So, I still left the `-1` in the `pass2`. Do you think it is OK?

**Comment by yanring** [comment]
Hi @timmoon10 @phu0ngng, could you help take another look at this? We intend to incorporate this optimization into mcore v0.11 (6th Feb). Thanks a lot!

File: transformer_engine/pytorch/triton/permutation.py (hunk 7 above)
Comment (phu0ngng) [reply transformer_engine/pytorch/triton/permutation.py]: Sounds good to me.

**Review by phu0ngng — APPROVED** [review]

File: transformer_engine/pytorch/permutation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/permutation.py]: We still have some places in the tests that use "indice", but it's less important since it's internal.

File: tests/pytorch/test_permutation.py [hunk 9]
```diff
@@ -480,6 +1049,57 @@ def test_permutation_topk1_no_probs(
...
+        BENCHMARK=BENCHMARK,
+    )
+
+
+@pytest.mark.parametrize("te_dtype", _te_dtypes)
+@pytest.mark.parametrize("num_tokens", [4096])
+@pytest.mark.parametrize("num_expert", [8, 16])
+@pytest.mark.parametrize("tp_size", [1, 2, 8])
+@pytest.mark.parametrize("hidden_size", [4096])
+def test_chunk_permuation(
```
Comment (timmoon10) [inline tests/pytorch/test_permutation.py:1084]: Nit:
```suggestion
def test_chunk_permutation(
```

File: tests/pytorch/test_permutation.py [hunk 10]
```diff
@@ -181,21 +181,27 @@ def pytorch_sort_chunks_by_idxs(
...
+        if te_dtype == tex.DType.kBFloat16:
+            return dict(rtol=2.0e-2, atol=1.0e-5)
+        if te_dtype == tex.DType.kFloat8E5M2 or te_dtype == tex.DType.kFloat8E4M3:
+            return dict(rtol=2.0e-1, atol=1.0e-1)
+        raise ValueError(f"Unsuppored dtype ({te_dtype})")
+    else:
+        if te_dtype == tex.DType.kFloat8E5M2 or te_dtype == tex.DType.kFloat8E4M3:
+            return dict(rtol=2.0e-1, atol=1.0e-1)
+        else:
+            return dict(rtol=0.0, atol=0.0)
```
Comment (timmoon10) [inline tests/pytorch/test_permutation.py:204]: I think this change makes this function much less intuitive. With `has_tolerance=False`, why do we expect bitwise accuracy with FP32 but not with FP8? It makes sense if you know the details of the MoE impl, but why should you need to know that in a simple helper function like this?

Also, based on your comments I think using dtype tolerances is fine for these tests. I think we can just revert this function.

File: tests/pytorch/test_permutation.py (hunk 4 above)
Comment (timmoon10) [reply tests/pytorch/test_permutation.py:594]: If we're doing reductions in unpermutation then it makes sense to use the dtype tolerances.

**Review by timmoon10 — APPROVED** [review]
Overall LGTM. My suggestions are stylistic.

File: transformer_engine/pytorch/permutation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hxbai) [reply transformer_engine/pytorch/permutation.py]: Thanks. I fixed these names.

File: tests/pytorch/test_permutation.py (hunk 9 above)
Comment (hxbai) [reply tests/pytorch/test_permutation.py]: Fixed.

File: tests/pytorch/test_permutation.py (hunk 10 above)
Comment (hxbai) [reply tests/pytorch/test_permutation.py]: OK. Just reverted.
